SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,193 @@
+                throw new StorageException("Failed to open database connection.", e);
+            }
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to open database connection, result={result}\n\t{errorMessage}");
             }
         }
 
-        public async Task CreateTableAsync<T>() where T : new()
+        public void Dispose()
         {
-            try
+            if (_db == null)
             {
-                // In SQLite-net 1.5 return type was changed.
-                // Using reflection to accept newer library version.
-                var task = (Task)_dbConnection.GetType()
-                    .GetMethod("CreateTableAsync", new[] { typeof(CreateFlags) })
-                    .MakeGenericMethod(typeof(T))
-                    .Invoke(_dbConnection, new object[] { CreateFlags.None });
-                await task.ConfigureAwait(false);
-            }
-            catch (SQLiteException e)
-            {
-                throw ToStorageException(e);
+                AppCenterLog.Error(AppCenterLog.LogTag, "Trying to dispose storage adapter while database is null.");
+                return;
             }
+            raw.sqlite3_close(_db);
+            _db.Dispose();
         }
 
-        public async Task<List<T>> GetAsync<T>(Expression<Func<T, bool>> pred, int limit) where T : new()
+        private void BindParameter(sqlite3_stmt stmt, int index, object value)
         {
-            try
+            int result;
+            if (value is string)
+            {
+                result = raw.sqlite3_bind_text(stmt, index, (string)value);
+            }
+            else if (value is int)
             {
-                var table = _dbConnection.Table<T>();
-                return await table.Where(pred).Take(limit).ToListAsync().ConfigureAwait(false);
+                result = raw.sqlite3_bind_int(stmt, index, (int)value);
             }
-            catch (SQLiteException e)
+            else if (value is long)
             {
-                throw ToStorageException(e);
+                result = raw.sqlite3_bind_int64(stmt, index, (long)value);
+            }
+            else
+            {
+                throw new NotSupportedException($"Type {value.GetType().FullName} not supported.");
+            }
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to bind {index} parameter, result={result}\n\t{errorMessage}");
             }
         }
 
-        public Task<int> CountAsync<T>(Expression<Func<T, bool>> pred) where T : new()
+        private void BindParameters(sqlite3_stmt stmt, IList<object> values)
         {
-            var table = _dbConnection.Table<T>();
-            return table.Where(pred).CountAsync();
+            for (int i = 0; i < values?.Count; i++)
```

Comment:
`values?.Count` will return `null` when values is `null`, then comparing an integer with `null` will throw.

Reply:
I was wrong. In C# comparing `null` with a number will result in `false`. So we are good.